File: pfun_cma_model/routes/dexcom.py
"""
Dexcom API routes.
"""
import os
import logging
from fastapi import APIRouter, Request, Depends
from fastapi.responses import RedirectResponse
import httpx
import pkce

# Configure logging
logger = logging.getLogger(__name__)

# Create a new router
router = APIRouter()

# Dexcom API configuration
DEXCOM_CLIENT_ID = os.getenv("DEXCOM_CLIENT_ID")
DEXCOM_CLIENT_SECRET = os.getenv("DEXCOM_CLIENT_SECRET")
DEXCOM_REDIRECT_URI = os.getenv("DEXCOM_REDIRECT_URI")
DEXCOM_BASE_URL = "https://sandbox-api.dexcom.com" # or "https://api.dexcom.com" for production

@router.get("/dexcom/auth")
async def dexcom_auth(request: Request):
    """
    Redirects the user to the Dexcom login page to initiate the OAuth 2.0 flow.
    """
    code_verifier, code_challenge = pkce.generate_pkce_pair()
    request.session['code_verifier'] = code_verifier

    params = {
        "client_id": DEXCOM_CLIENT_ID,
        "redirect_uri": DEXCOM_REDIRECT_URI,
        "response_type": "code",
        "scope": "offline_access",
        "state": "some_random_state_string", # Should be a random string
        "code_challenge": code_challenge,
        "code_challenge_method": "S256",
    }

    auth_url = f"{DEXCOM_BASE_URL}/v2/oauth2/login?{httpx.URL(params).query.decode()}"
    return RedirectResponse(url=auth_url)

@router.get("/dexcom/auth/callback")
async def dexcom_auth_callback(request: Request):
    """
    Handles the callback from Dexcom after the user has authenticated.
    """
    code = request.query_params.get("code")
    code_verifier = request.session.pop('code_verifier', None)

    if not code or not code_verifier:
        return {"error": "Invalid request"}

    data = {
        "grant_type": "authorization_code",
        "code": code,
        "redirect_uri": DEXCOM_REDIRECT_URI,
        "code_verifier": code_verifier,
        "client_id": DEXCOM_CLIENT_ID,
        "client_secret": DEXCOM_CLIENT_SECRET,
    }

    async with httpx.AsyncClient() as client:
        response = await client.post(f"{DEXCOM_BASE_URL}/v2/oauth2/token", data=data)

    if response.status_code != 200:
        return {"error": "Failed to fetch access token"}

    token_data = response.json()
    request.session['dexcom_access_token'] = token_data['access_token']
    request.session['dexcom_refresh_token'] = token_data['refresh_token']

    return RedirectResponse(url="/demo/dexcom")

@router.api_route("/dexcom/api/{path:path}", methods=["GET", "POST", "PUT", "DELETE"])
async def dexcom_api_proxy(request: Request, path: str):
    """
    A proxy for the Dexcom API.
    """
    access_token = request.session.get('dexcom_access_token')
    if not access_token:
        return {"error": "Not authenticated"}, 401

    headers = {
        "Authorization": f"Bearer {access_token}",
        "Content-Type": request.headers.get("Content-Type"),
    }

    url = f"{DEXCOM_BASE_URL}/{path}"

    async with httpx.AsyncClient() as client:
        if request.method == "GET":
            response = await client.get(url, headers=headers, params=request.query_params)
        elif request.method == "POST":
            response = await client.post(url, headers=headers, json=await request.json())
        # Add other methods as needed
        else:
            return {"error": "Method not supported"}, 405

    return response.json(), response.status_code

File: packages/api/src/routes/dexcom.ts
/**
 * Dexcom routes: /dexcom/*
 */

import { Hono } from "hono";
import type { AppConfig } from "../config.js";

// Simple in-memory session store (production would use Redis)
const sessions = new Map<string, Record<string, string>>();

function getSessionId(c: { req: { header: (name: string) => string | undefined } }): string {
  return c.req.header("x-session-id") ?? "default";
}

export function createDexcomRoutes(config: AppConfig): Hono {
  const app = new Hono();

  const DEXCOM_BASE = "https://sandbox-api.dexcom.com";

  // GET /dexcom/test
  app.get("/test", (c) => {
    return c.json({ message: "Dexcom router is working" });
  });

  // POST /dexcom/token
  app.post("/token", async (c) => {
    const body = await c.req.json();
    const { code, redirect_uri } = body;

    if (!code) {
      return c.json({ detail: "Missing authorization code" }, 400);
    }

    try {
      const tokenRes = await fetch(`${DEXCOM_BASE}/v2/oauth2/token`, {
        method: "POST",
        headers: { "Content-Type": "application/x-www-form-urlencoded" },
        body: new URLSearchParams({
          client_id: config.dexcomClientId,
          client_secret: config.dexcomClientSecret,
          code,
          grant_type: "authorization_code",
          redirect_uri: redirect_uri ?? config.dexcomRedirectUri,
        }),
      });

      if (!tokenRes.ok) {
        const errData = await tokenRes.json().catch(() => ({}));
        return c.json({ detail: errData }, (tokenRes.status >= 400 && tokenRes.status < 600 ? tokenRes.status : 500) as 400);
      }

      const data = (await tokenRes.json()) as {
        access_token: string;
        refresh_token: string;
      };

      // Store tokens in session
      const sessionId = getSessionId(c);
      const session = sessions.get(sessionId) ?? {};
      session.dexcom_access_token = data.access_token;
      session.dexcom_refresh_token = data.refresh_token;
      sessions.set(sessionId, session);

      return c.json(data);
    } catch (err) {
      return c.json({ detail: "Token exchange failed", error: String(err) }, 500);
    }
  });

  // GET /dexcom/auth/callback
  app.get("/auth/callback", (c) => {
    const code = c.req.query("code");
    if (!code) {
      return c.json({ detail: "Missing code parameter" }, 400);
    }

    const sessionId = getSessionId(c);
    const session = sessions.get(sessionId) ?? {};
    session.dexcom_auth_code = code;
    sessions.set(sessionId, session);

    return c.redirect("/demo/dexcom");
  });

  // GET /dexcom/users/self/egvs
  app.get("/users/self/egvs", async (c) => {
    const sessionId = getSessionId(c);
    const session = sessions.get(sessionId);
    const accessToken = session?.dexcom_access_token;

    if (!accessToken) {
      return c.json({ detail: "No access token. Please authenticate first." }, 401);
    }

    const startDate = c.req.query("startDate") ?? "";
    const endDate = c.req.query("endDate") ?? "";

    try {
      const authHeader = "Bearer " + accessToken;
      const res = await fetch(
        `${DEXCOM_BASE}/v2/users/self/egvs?startDate=${startDate}&endDate=${endDate}`,
        { headers: { Authorization: authHeader } },
      );
      const data = await res.json();
      return c.json(data, 200);
    } catch (err) {
      return c.json({ detail: "Failed to fetch EGVs", error: String(err) }, 500);
    }
  });

  // GET /dexcom/users/self/devices
  app.get("/users/self/devices", async (c) => {
    const sessionId = getSessionId(c);
    const session = sessions.get(sessionId);
    const accessToken = session?.dexcom_access_token;

    if (!accessToken) {
      return c.json({ detail: "No access token. Please authenticate first." }, 401);
    }

    try {
      const authHeader = "Bearer " + accessToken;
      const res = await fetch(`${DEXCOM_BASE}/v2/users/self/devices`, {
        headers: { Authorization: authHeader },
      });
      const data = await res.json();
      return c.json(data, 200);
    } catch (err) {
      return c.json({ detail: "Failed to fetch devices", error: String(err) }, 500);
    }
  });

  return app;
}
